bip-framework-libraries/src/main/java/gov/va/bip/framework/security/jwt/JwtParser.java

```
@@ -45,27 +48,84 @@ public JwtParser(final JwtAuthenticationProperties properties) {
+		if (claims == null && StringUtils.isNotBlank(jwtAuthenticationProperties.getIssuer())
+				&& StringUtils.isNotBlank(jwtAuthenticationProperties.getSecret())) {
+			// We will sign our JWT with our ApiKey secret
+			final Key signingKey = createSigningKey(jwtAuthenticationProperties.getSecret());
 
-		claims = Jwts.parser().setSigningKey(signingKey).requireIssuer(jwtAuthenticationProperties.getIssuer())
-				.parseClaimsJws(token).getBody();
+			claims = Jwts.parser().setSigningKey(signingKey).requireIssuer(jwtAuthenticationProperties.getIssuer())
+					.parseClaimsJws(token).getBody();
+		}
 
 		final long elapsedTime = System.currentTimeMillis() - startTime;
 
 		LOGGER.debug("Time elapsed to parse JWT token {}{}{}", "[", elapsedTime / NUMBER_OF_MILLIS_N_A_SECOND,
 				" secs]");
 
-		return getPersonFrom(claims);
+		if (claims != null) {
+			return getPersonFrom(claims);
+		} else {
+			return null;
+		}
+	}
 
+	/**
+	 * Parses the JWT key pairs.
+	 *
+	 * @param token
+	 *            the token
+	 * @param claims
+	 *            the claims
+	 * @return the claims
+	 */
+	private Claims parseJwtKeyPairs(final String token) {
+		
+		Claims claims = null;
+		final List<JwtKeyPairs> jwtKeyPairs = jwtAuthenticationProperties.getKeyPairs();
+		if (jwtKeyPairs != null && !jwtKeyPairs.isEmpty()) {
+			for (int i = 0; i < jwtKeyPairs.size(); i++) {
+				final JwtKeyPairs jwtKeyPair = jwtKeyPairs.get(i);
+				// We will sign our JWT with our ApiKey secret
+				Key signingKey = createSigningKey(jwtKeyPair.getSecret());
+				try {
+					claims = Jwts.parser().setSigningKey(signingKey).requireIssuer(jwtKeyPair.getIssuer())
+							.parseClaimsJws(token).getBody();
+					LOGGER.debug("claims {}", claims);
+				} catch (final Exception e) {
+					LOGGER.error("Exception parsing JWT token {}", e);
+					if (i == jwtKeyPairs.size() - 1) {
+						LOGGER.error("Rethrowing Error");
+						throw e;
+					} else {
+						LOGGER.warn("Trying next JWT Key Pair");
+					}
```

This catch seems to swallow and log the exception if its not the last KeyPair in the list, but then actually rethrow the exception if it does happen to be the last KeyPair in the list.

I feel like this should be more consistent with how it handles these exceptions. If we really don't care about the exception and want to try the next KeyPair in the list, then why rethrow the exception on the last one? Wouldn't it be better to try and default back to the original _jwtAuthenticationProperties.getIssuer()_ and _jwtAuthenticationProperties.getSecret()_ in the logic?

I'd vote for this: 
Option 1: An exception that occurs here should just be rethrown up immediately, even if there are other KeyPairs to check
OR
Option 2: Swallow and log all of the exceptions, and just let the code continue on and default to the original Issuer/Secret keys

---

Option 1 is NOT an option at all. With multiple consumers to support as an objective. you can't throw exception immediately, else you will likely never ever let's say Consumer 2 invoke the protected end points. With Option 2, problem is that you are letting the services configure multiple key pairs, try those for the token, but then also relying on the fallback single key pair option. I don't want that in consistent behavior. The reason I am throwing exception for last key pair is the fact that the system tried all configured multiple key pair options but they failed. and I want to get that exception back vs trying for default (properties has default values in the bean) and failing again.

---

I see what you are getting at now, my apologies. If we are relying on the exception to tell us that it's not the correct KeyPair, then can we catch the specific exception types instead of the general `Exception`? My problem from the start was that, if something else really weird happens (i.e. the `token` that is passed in is null, or the Jwts.parser() call hits an exception), then that should be handled differently.

`parseClaimsJwt(...)` throws UnsupportedJwtException, MalformedJwtException, SignatureException, ExpiredJwtException, and IllegalArgumentException. I feel like maybe some of these should be handled differently, i.e. MalformedJwtException or IllegalArgumentExceptions should maybe rethrow immediately, since it points to a bigger issue than just "the KeyPair didn't match"

---

Thanks for your feedback. It's a good one, however there are multiple calls being made including signing key + require issuer + parse claims JWS.  General exception will log it anyway plus without trying all pairs, I am hesitant to throw the exception back immediately for the misconfiguration of a consumer pair. Pick the poison, so inclined to leave it as is.